src/main/java/Duke.java

```diff
@@ -1,10 +1,133 @@
+import java.util.Scanner;
+
 public class Duke {
+    private static final String SAVE_FILE_PATH =
+            System.getProperty("user.home") + File.separator + ".duke" + File.separator + "tasks.txt";
+    private static final String BYE = "bye";
+
+    private final Storage storage;
+    private final Ui ui;
+    private final TaskList tasks;
+
+    public Duke(String filePath) {
+        this.ui = new Ui();
+        this.storage = new Storage(filePath);
+        this.tasks = new TaskList();
+        try {
+            this.tasks.addAllTasks(this.storage.load());
+        } catch (IOException e) {
+            this.ui.showLoadingError();
+        }
+    }
+
+    /**
+     * Initializes an instance of Duke, and runs it.
+     * @param args The command line args passed to the program
+     */
     public static void main(String[] args) {
-        String logo = " ____        _        \n"
-                + "|  _ \\ _   _| | _____ \n"
-                + "| | | | | | | |/ / _ \\\n"
-                + "| |_| | |_| |   <  __/\n"
-                + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+        new Duke(SAVE_FILE_PATH).run();
+    }
+
+    /**
+     * Executes an instance of the Duke chatbot. While the user does not input a goodbye command, Duke interprets
+     * different commands passed to it and performs different actions, including but not limited to the following:
+     * - store a task to be done
+     * - mark a task as done
+     * - list the tasks to be done
+     * - find tasks that match a search term
+     * - remove a task from the list of tasks
+     */
+    private void run() {
+        try (Scanner scanner = new Scanner(System.in)) {
+            this.ui.displayGreeting();
+
+            if (this.storage.hasSavedTasks()) {
+                this.ui.displayGreetingReminder(this.tasks.tasksCount());
+            }
+
+            System.out.print("> ");
+            while (scanner.hasNextLine()) {
+                String input = scanner.nextLine();
+
+                if (input.equalsIgnoreCase(BYE)) {
+                    this.storage.saveTasks(this.tasks.getTasks());
+                    break;
+                } else {
+                    this.handleCommand(input);
+                }
+                System.out.print("> ");
+            }
+
+            this.ui.displayGoodbye();
+        } catch (IOException e) {
+            this.ui.displayMessages(e.getMessage());
+        }
+
+    }
+
+    /**
+     * Interprets a command that was passed to Duke, and performs a corresponding action based on it. If the command or
+     * its arguments were malformed, display an error to the user.
+     *
+     * @param input The String containing the command, as well as its arguments
+     */
+    private void handleCommand(String input) {
+        try {
+            String[] tokens = input.split(" ");
+            String command = tokens[0].toLowerCase();
+            switch (command) {
```

Comment:
Should logical units within a block be placed together? Maybe separating them by 1 line could improve readability